### PR TITLE
Catch encoding errors

### PR DESCRIPTION
**Pull request by CiaccoDavide**
Try to encode the dweet while trying to compress/decompress it
(maybe it should be checked every time the code changes?)

### CODE REVIEW

**Comment by CiaccoDavide** [comment]
I don't think this solves the linked issue, yet...

pasting the following code in dwitter beta, breaks it:
```js
eval(unescape(escape`𭱩𭁨𚁸𚑸𛡦👦𪑬𫁔𩑸𭀬𩡯𫡴🐧𞑩𫠢𙰬𩠨�됽똶�𛀳𝠷𛀸𝀴𚐬𩡯𫡴🐧𜰮𝑩𫠢𙰬𩠨�됾떋�𛀷𜰰𛀸𝠰𚐬𩡯𫡴🐧𜀮𞁩𫠢𙰬𩠨�됽뒀�𛀹𜐰𛀷𝰰𚐠`.replace(/u../g,'')))
```

but not on the standard website, creating a dweet with this code on "standard dwitter" doesn't break anything: https://beta.dwitter.net/d/24256

**Comment by lionleaf** [comment]
Seems like a reasonable approach, test it and fall back. But really there should be a way to actually run those dweets; any idea why it fails? Are there certain unicode characters that just break URIencoding?

**Comment by CiaccoDavide** [comment]
> Are there certain unicode characters that just break URIencoding?

Apparently, yeah!
I tried to test `encodeURIComponent` on 900k characters: https://ciaccodavi.de/storage/testing/dwitter/

Then I found this answer on stackoverflow: https://stackoverflow.com/a/17109094/2104882

Seems like all the unicode characters between `\ud800` and `\udfff` are not supported (why though...) by `encodeURIComponent`...

So I guess that the problem is not the 😶 emoji itself, but the unicode characters that it's converted to during the dweet compression!

**Comment by CiaccoDavide** [comment]
Ok, got it...compression by unicode surrogates pairing (two characters into one) works by joining two unicode characters included inside the interval between `0xD800` and `0xDFFF`.

The first element of the pair has to be taken from the first half of the interval, while the second one is one of the last 1024 values (from `0xDC00` to `0xDFFF`).

This works well for compressing most characters (since we mostly write code using ASCII), but when the compression algorithm encounters charcodes greater than 1024 can't correctly build surrogate pairs (that's why the emojis disappear and are being replaced by other characters).

**Comment by CiaccoDavide** [comment]
Ok, I thought abount it and I don't really see a way out of it...
The only solutions that I found were:
 - change the compressor to something that extracts all the unsupported characters (charCodes greater than 1024) and uses them as variables (example: ```js eval(unescape(escape`compressed string`).replace(/u../g,''))}{A['😶','🐧']```), but I don't think it's a good idea because it could produce lengthier code and it's a change that should be done by the user
 - add 4096 to all charcodes above 1024, but then the decompression code should have an other `replace` (`.replace(/u[^E]./g,'').replace(/uE./g,'uD')`) and it doesn't always work

So...since some data is lost anyway when you try to compress code with charcodes greater than 1024 inside I think the best way to tackle this should be showing an error when it crashes and maybe a warning when it doesn't crash but data is lost after compression/decompression (see the videos above, the emojis are dissected into two other characters)

What do you think? @lionleaf @sigvef

**Comment by CiaccoDavide** [comment]
Currently this PR only adds a try/catch that displays the error `An error occurred while trying to compress the code`:

https://user-images.githubusercontent.com/3226096/139735887-91870369-76a8-4dfe-9f7b-a9a7f5e6e702.mp4

**Comment by CiaccoDavide** [comment]
Added a warning for when the code is compressed but data could be lost:

https://user-images.githubusercontent.com/3226096/139740897-325f3503-07dc-4efe-a4f7-f478bad08937.mp4

**Comment by lionleaf** [comment]
So I think the long term solution here would be to not rely on uriEncoding at all. I've been wanting to change the way the frontend displays dweets anyway, so that's likely to happen at some point.

But for now I think this is a great temporary solution!

I guess dweets have been using invalid unicode characters for a while huh. 

Will review properly soon.

**Comment by CiaccoDavide** [comment]
> So I think the long term solution here would be to not rely on uriEncoding at all

yeah 😅 


> I've been wanting to change the way the frontend displays dweets anyway, so that's likely to happen at some point

That would be awesome, were you thinking about getting rid of the iframes altogheter (just using canvases) or something else?

**Comment by CiaccoDavide** [comment]
I think I got something here!

I rewrote both the compressor and the decompressor, and it seems to work well...it still needs a bit of testing but it solves the unicode problem (by not compressing characters with a code greater than `CF` (`207`) )

I placed both functions inside [utils.ts](https://github.com/dwitter-net/dwitter-frontend/blob/d62e539b83a3d0e5f44d4748e86cbc4d8b59ea5c/src/utils.ts)

**Comment by lionleaf** [comment]
Very cool! 

Does this impact decompression of existing dweets in any way? As the main utility of that button was to make it easier to read other people's compressed code

File: src/utils.ts [hunk 1]
```diff
@@ -1,2 +1,88 @@
 export const getDweetLength = (code: string) =>
   [...code.replace(/\r\n/g, '\n')].length;
+
+// const getSurrogatePair = (firstCharCode: number, secondCharCode: number) => {
```
Comment (lionleaf) [inline src/utils.ts:4]: Left-over code?

**Review by lionleaf — APPROVED** [review]
LGTM and runs locally :) Feel free to press the big merge button once you remove the commented-out function (unless you meant for it to be there for reference?).

Thanks for the effort put into this!

**Comment by CiaccoDavide** [comment]
> unless you meant for it to be there for reference?

Yeah, I was going to leave it for reference but I guess I could write a wiki page on the different types of compressors, if that's ok (I think that could be interesting to read for someone who doesn't know how they work)
